src/bin/cargo/cli.rs

```diff
@@ -261,6 +261,18 @@ fn expand_aliases(
             }
             (None, None) => {}
             (_, Some(mut alias)) => {
+                // Check if this alias is shadowing an external subcommand
+                // (binary of the form `cargo-<subcommand>`)
+                // Currently this is only a warning, but after a transition period this will become
+                // a hard error.
+                if let Some(path) = super::find_external_subcommand(config, cmd) {
+                    config.shell().warn(format!(
+                        "user-defined alias `{}` is shadowing an external subcommand found at: `{}`",
```

Comment:
Can this include additional text explaining that this will become an error in the future?  And some detail explaining why it is being changed (perhaps just a link to https://github.com/rust-lang/cargo/issues/10049 will be sufficient).  Maybe something like this?

```
This was previously accepted but is being phased out; it will become a hard error in a future release.
For more information, see issue #10049 <https://github.com/rust-lang/cargo/issues/10049>.
```